File: core/src/net/bossdragon/system/view/render/anim/KeyFrameAnimationsMap.java
package net.bossdragon.system.view.render.anim;

import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.utils.Array;
import net.bossdragon.util.collections.RandomAccessMap;
import net.bossdragon.util.collections.tuples.TripleList;
import net.bossdragon.util.data.Triple;

public class KeyFrameAnimationsMap {
    private final RandomAccessMap<String, Animation> _map;


    public KeyFrameAnimationsMap() {
        _map = new RandomAccessMap<>();
    }

    public KeyFrameAnimationsMap(int initialSize) {
        _map = new RandomAccessMap<>(initialSize);
    }

    public void put(String name, Animation anim) {
        _map.put(name, anim);
    }

    public Animation get(String name) {
        return _map.get(name);
    }

    public int size() {
        return _map.size();
    }

    public static KeyFrameAnimationsMap create(TripleList<String, Float, Integer> anims, Array<Array<TextureRegion>> frames) {
        final KeyFrameAnimationsMap map = new KeyFrameAnimationsMap(anims.size());

        for (Triple<String, Float, Integer> triple : anims) {
            final String name = triple.Item1;
            final float frameDuration = triple.Item2;
            final Array<TextureRegion> keyFrames = frames.get(triple.Item3);
            final Animation anim = new Animation(frameDuration, keyFrames);

            map.put(name, anim);
        }

        return map;
    }

    public static KeyFrameAnimationsMap create(float frameDuration, Array<TextureRegion> frames) {
        KeyFrameAnimationsMap map = new KeyFrameAnimationsMap(1);
        Animation anim = new Animation(frameDuration, frames);
        map.put("default", anim);
        return map;
    }
}

File: core/src/net/bossdragon/system/view/render/anim/KeyFrameAnimationsMap.kt
package net.bossdragon.system.view.render.anim

import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.TextureRegion
import net.bossdragon.util.collections.RandomAccessMap
import net.bossdragon.util.collections.tuples.TripleList

class KeyFrameAnimationsMap {
    private val _map: RandomAccessMap<String, Animation<*>>


    constructor() {
        _map = RandomAccessMap<String, Animation<*>>()
    }

    constructor(initialSize: Int) {
        _map = RandomAccessMap<String, Animation<*>>(initialSize)
    }

    fun put(name: String, animation: Animation<*>) {
        _map.put(name, animation)
    }

    operator fun get(name: String): Animation<*>? {
        return _map.get(name)
    }

    operator fun get(index: Int): Animation<*> {
        return _map.get(index)
    }

    fun size(): Int {
        return _map.size
    }

    companion object {

        fun create(animations: TripleList<String, Float, Int>, frames: Array<Array<TextureRegion>>): KeyFrameAnimationsMap {
            val map = KeyFrameAnimationsMap(animations.size())

            for (triple in animations) {
                val name = triple.Item1
                val frameDuration = triple.Item2
                val keyFrames = frames[triple.Item3]
                val animation = Animation(frameDuration, *keyFrames)

                map.put(name, animation)
            }

            return map
        }

        fun create(frameDuration: Float, frames: Array<TextureRegion>): KeyFrameAnimationsMap {
            val map = KeyFrameAnimationsMap(1)

            val animation = Animation(frameDuration, *frames)
            map.put("default", animation)
            return map
        }
    }
}
